### ISSUE TITLE
Publish to Maven Central

### ISSUE THREAD

**Issue by bowbahdoe** [bowbahdoe/magic-bean#8]
What we want is probably a github action that automatically publishes on a tagged release.

(Some people just aren't comfortable with Jitpack for whatever reason)

**Comment by chabala** [comment]
I'm glad you opened an issue. I'll move my link here: https://maven.apache.org/repository/guide-central-repository-upload.html

> a github action that automatically publishes on a tagged release.

That's a big goal for a first step. Better to get your project in shape to release to Central and get a release out before trying to automate it.

Do you own `mccue.dev`? If not, you should either try to buy it, or choose new coordinates that you own. It's not optional: https://central.sonatype.org/publish/requirements/coordinates/

Once you determine which coordinates to use, you can setup an account with Sonatype to get permission to publish to them: https://central.sonatype.org/publish/publish-guide/#initial-setup . This is the mechanism that keeps coordinates unique in Java, and not a free for all like NPM.

**Comment by bowbahdoe** [comment]
I do own it. (I just messed up the https on the silly site i put on it and didn't fix it)

If I didn't own it Jitpack wouldn't publish under dev.mccue - it would publish under com.github.bowbahdoe

**Comment by bowbahdoe** [comment]
One thing I don't quite understand though - other than requiring users to add another entry to their repositories I don't see why I would prefer maven central over jitpack.

**Comment by chabala** [comment]
The reason you don't have to add central to the repository list to use it is because it's the canonical repository. It's the gold standard for publishing artifacts. It will never go away, artifacts never disappear from it, you can depend on it being there, forever.

Unlike Bintray/JCenter, which caught a lot of people by surprise when it announced plans to shutdown. JitPack is even more suspect, it's not clear who's running it or if it will disappear tomorrow.

If you want people to take your project seriously and use it in their builds, it should be on Central.

**Comment by bowbahdoe** [comment]
Making a checklist here to keep myself accountable. 


- [X] Create Jira account
- [X] Create new project ticket
- [X] Project Name, Description and URL
- [X] License Information
- [X] Developer Information
- [X] SCM Information
- [X] Figure out how the heck to use gpg and where to put my private key
- [X] Build and initial publish
- [X] Add a github action to build
- [X] Add a github action to publish

**Comment by chabala** [comment]
Just a nudge, asking for the project namespace via JIRA is the part that takes the longest, because a human at Sonatype actually works on it, but it should only take a few days at the most. Once you're approved for `dev.mccue` you'll be able to publish any artifacts under that namespace in any project without needing to contact them again.

**Comment by bowbahdoe** [comment]
Okay made ticket, going to tackle some of the metadata tonight

**Comment by bowbahdoe** [comment]
Provided the TXT proof and re-opened ticket. now just waiting on them

**Comment by bowbahdoe** [comment]
Got ticked approved and filled in most info.

**Comment by chabala** [comment]
Nice job, looks like it's even more automated now than when I used it: https://issues.sonatype.org/browse/OSSRH-78053

**Comment by bowbahdoe** [comment]
```
Error:  Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8:deploy (injected-nexus-deploy) on project magic-bean: Execution injected-nexus-deploy of goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8:deploy failed: An API incompatibility was encountered while executing org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8:deploy: java.lang.ExceptionInInitializerError: null
Error:  -----------------------------------------------------
Error:  realm =    extension>org.sonatype.plugins:nexus-staging-maven-plugin:1.6.8
Error:  strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
Error:  urls[0] = file:/home/runner/.m2/repository/org/sonatype/plugins/nexus-staging-maven-plugin/1.6.8/nexus-staging-maven-plugin-1.6.8.jar
Error:  urls[1] = file:/home/runner/.m2/repository/org/sonatype/nexus/maven/nexus-common/1.6.8/nexus-common-1.6.8.jar
Error:  urls[2] = file:/home/runner/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.4/plexus-sec-dispatcher-1.4.jar
Error:  urls[3] = file:/home/runner/.m2/repository/org/sonatype/plexus/plexus-cipher/1.7/plexus-cipher-1.7.jar
Error:  urls[4] = file:/home/runner/.m2/repository/com/google/guava/guava/14.0.1/guava-14.0.1.jar
```

Every step of this has been the biggest PITA *except* for proving I owned the domain.

**Comment by bowbahdoe** [comment]
https://issues.sonatype.org/browse/OSSRH-66257

Absolutely busted. 

Without this random gist there would have been no way I could have figured out as much as I have.

https://gist.github.com/sualeh/ae78dc16123899d7942bc38baba5203c

> It's the gold standard for publishing artifacts

I believe that it is the gold standard for **hosting** published artifacts, but damn this is a gate well kept.

**Comment by chabala** [comment]
Ah, yeah, never seen that myself. Maybe you wanna prototype this work on a branch instead of main.

**Comment by bowbahdoe** [comment]
🤷 Already done. I didn't really know how to test a release except on main.

On central, just tested in a brand new project. Action should be set up for any future releases.

**Comment by bowbahdoe** [comment]
@chabala Regardless, thank you for your help. I'm gonna close this issue as soon as the repo appears on search.

**Comment by bowbahdoe** [comment]
<img width="636" alt="image" src="https://user-images.githubusercontent.com/5004262/153438485-6e1a38dc-1263-4aac-82b8-397e87b5316a.png">
